### PR TITLE
allow multistream to have zero rtt stream opening

### PR DESCRIPTION
**Pull request by whyrusleeping**
This is probably WIP, i need to look over it again to make sure. But i need to work on the GC stuff before i get back to this. 

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

File: Godeps/_workspace/src/github.com/whyrusleeping/go-multistream/lazy.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		l.rerr = fmt.Errorf("multistream protocol mismatch ( %s != %s )", tok, ProtocolID)
+		return l.rerr
+	}
+
+	// read protocol
+	tok, err = ReadNextToken(l.con)
+	if err != nil {
+		l.rerr = err
+		return err
+	}
```
Comment (jbenet) [inline Godeps/_workspace/src/github.com/whyrusleeping/go-multistream/lazy.go]: to me this seems like a special case of the muxer: writing two headers back to back. loses generality. it's fine for this use case, but i'd almost wrap this in two lazyConns, one for each protocol ID.

File: Godeps/_workspace/src/github.com/whyrusleeping/go-multistream/lazy.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	werr   error
+
+	proto string
+	con   io.ReadWriteCloser
+}
+
+func (l *lazyConn) Read(b []byte) (int, error) {
+	if !l.rhandshake {
+		go l.writeHandshake()
+		err := l.readHandshake()
```
Comment (jbenet) [inline Godeps/_workspace/src/github.com/whyrusleeping/go-multistream/lazy.go]: i like this idea :+1:

**Comment by jbenet** [comment]
2 comments, LGTM

**Comment by whyrusleeping** [comment]
alright, RFM?

**Comment by jbenet** [comment]
@whyrusleeping i think it's worth layering the two different lazyConns. it will make using other multistreams way easier.

**Comment by whyrusleeping** [comment]
should we also then break up the muxer?

**Comment by jbenet** [comment]
hmm the muxer is probably fine. i think `LazyConn` is more general then multistream-select

**Comment by jbenet** [comment]
@whyrusleeping LGTM!

**Comment by jbenet** [comment]
let's wait for tests

**Comment by whyrusleeping** [comment]
@jbenet  this looks like that peerstream bug again: https://circleci.com/gh/ipfs/go-ipfs/602

You fixed that right?

**Comment by jbenet** [comment]
yep this is fixed in master

**Comment by whyrusleeping** [comment]
@jbenet this should be RFM

**Pull request merged**